### ISSUE TITLE
possible race condition in template system

### ISSUE THREAD

**Issue by TankTheFrank** [astaxie/beego#2544]
1. What version of Go and beego are you using (`bee version`)?
go1.8
beego 1.8.0

2. What operating system and processor architecture are you using (`go env`)?
darwin (osx Sierra, 10.12.4) amd64

3. What did you do?
Created a new project and added the two template files from tutorial: https://beego.me/docs/mvc/view/view.md#another-approach

controller:
```
this.TplName = "blog/add.tpl"
this.Data["SomeVar"] = "SomeValue"
this.Data["Title"] = "Add"
```

add.tpl:
```
{{ template "layout_blog.tpl" . }}
{{ define "css" }}
        <link rel="stylesheet" href="/static/css/current.css">
{{ end}}


{{ define "content" }}
        <h2>{{ .Title }}</h2>
        <p> This is SomeVar: {{ .SomeVar }}</p>
{{ end }}

{{ define "js" }}
    <script src="/static/js/current.js"></script>
{{ end}}
```

layout_blog.tpl:
```
<!DOCTYPE html>
<html>
<head>
    <title>Lin Li</title>
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8">
    <link rel="stylesheet" href="http://netdna.bootstrapcdn.com/bootstrap/3.0.3/css/bootstrap.min.css">
    <link rel="stylesheet" href="http://netdna.bootstrapcdn.com/bootstrap/3.0.3/css/bootstrap-theme.min.css">
     {{ block "css" . }}{{ end }}
</head>
<body>

    <div class="container">
        {{ block "content" . }}{{ end }}
    </div>
    <script type="text/javascript" src="http://code.jquery.com/jquery-2.0.3.min.js"></script>
    <script src="http://netdna.bootstrapcdn.com/bootstrap/3.0.3/js/bootstrap.min.js"></script>
     {{ block "js" . }}{{ end }}
</body>
</html>
```

4. What did you expect to see?
Always the expected response (blocks filled with the provided content in add.tpl)

5. What did you see instead?
Either blank blocks or the correct response. Once every 10-15 requests (in dev mode, in production is even less) the correct response is rendered, otherwise the blocks are empty:

bad (all blocks empty response)
```
$ curl -v http://localhost:8080/
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 8080 (#0)
> GET / HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.51.0
> Accept: */*
>
< HTTP/1.1 200 OK
< Content-Length: 769
< Content-Type: text/html; charset=utf-8
< Server: beegoServer:1.8.0
< Date: Thu, 30 Mar 2017 20:56:31 GMT
<
<!DOCTYPE html>
<html>
<head>
    <title>Lin Li</title>
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8">
    <link rel="stylesheet" href="http://netdna.bootstrapcdn.com/bootstrap/3.0.3/css/bootstrap.min.css">
    <link rel="stylesheet" href="http://netdna.bootstrapcdn.com/bootstrap/3.0.3/css/bootstrap-theme.min.css">

        <link rel="stylesheet" href="/static/css/current.css">

</head>
<body>

    <div class="container">

    </div>
    <script type="text/javascript" src="http://code.jquery.com/jquery-2.0.3.min.js"></script>
    <script src="http://netdna.bootstrapcdn.com/bootstrap/3.0.3/js/bootstrap.min.js"></script>

</body>
</html>
```

bad response (content block is filled, js block is missing)
```
$ curl -v http://localhost:8080/
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 8080 (#0)
> GET / HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.51.0
> Accept: */*
>
< HTTP/1.1 200 OK
< Content-Length: 834
< Content-Type: text/html; charset=utf-8
< Server: beegoServer:1.8.0
< Date: Thu, 30 Mar 2017 20:56:29 GMT
<
<!DOCTYPE html>
<html>
<head>
    <title>Lin Li</title>
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8">
    <link rel="stylesheet" href="http://netdna.bootstrapcdn.com/bootstrap/3.0.3/css/bootstrap.min.css">
    <link rel="stylesheet" href="http://netdna.bootstrapcdn.com/bootstrap/3.0.3/css/bootstrap-theme.min.css">

        <link rel="stylesheet" href="/static/css/current.css">

</head>
<body>

    <div class="container">

        <h2>Add</h2>
        <p> This is SomeVar: SomeValue</p>

    </div>
    <script type="text/javascript" src="http://code.jquery.com/jquery-2.0.3.min.js"></script>
    <script src="http://netdna.bootstrapcdn.com/bootstrap/3.0.3/js/bootstrap.min.js"></script>

</body>
</html>
```

correct response:
```
$ curl -v http://localhost:8080/
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 8080 (#0)
> GET / HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.51.0
> Accept: */*
>
< HTTP/1.1 200 OK
< Content-Length: 885
< Content-Type: text/html; charset=utf-8
< Server: beegoServer:1.8.0
< Date: Thu, 30 Mar 2017 20:59:32 GMT
<
<!DOCTYPE html>
<html>
<head>
    <title>Lin Li</title>
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8">
    <link rel="stylesheet" href="http://netdna.bootstrapcdn.com/bootstrap/3.0.3/css/bootstrap.min.css">
    <link rel="stylesheet" href="http://netdna.bootstrapcdn.com/bootstrap/3.0.3/css/bootstrap-theme.min.css">

        <link rel="stylesheet" href="/static/css/current.css">

</head>
<body>

    <div class="container">

        <h2>Add</h2>
        <p> This is SomeVar: SomeValue</p>

    </div>
    <script type="text/javascript" src="http://code.jquery.com/jquery-2.0.3.min.js"></script>
    <script src="http://netdna.bootstrapcdn.com/bootstrap/3.0.3/js/bootstrap.min.js"></script>

    <script src="/static/js/current.js"></script>

</body>
</html>
```

**Comment by TheKLARKEN** [comment]
I'm also having the same issue! It seems to only be an issue in 1.8 (some people also say 1.7). I can get it to parse templates 10 times out of 10 if I change "runmode" to "prod" and only run the go application with  "go run" instead of "bee run." Otherwise, it works, as you've stated, very rarely, whether in prod or dev.

I'm not a GoPro (haha), but I could see a race condition causing this.

**Comment by TheKLARKEN** [comment]
Sorry to respond so much, but the issue seems to have been completely fixed in Go 1.8.1! I just tried it! My templates render perfectly!

**Comment by astaxie** [comment]
@TankTheFrank Could you try Go 1.8.1 now?

@TheKLARKEN  Thanks for the update
